### ISSUE TITLE
Rework integration with Moq to support generic methods

### ISSUE THREAD

**Issue by zvirja** [AutoFixture/AutoFixture#1139]
This issue is to track our activities regarding rework of Moq integration. Currently we are creating `Mock<T>` using reflection and then configure each method individually using reflection. Due to this approach we have a limitation and do not support generic methods, as we cannot configure all possible instantiations upfront.

We should change our integration approach and instead register AutoFixture as default value provider. This way we inject "passively" and are invoked by Moq when value is required. We are already following that way with NSubstitute and it works pretty well.

We haven't investigated yet whether it's possible, but @stakx promised to give it a look one day.

Let's postpone this activity till the moment we start our work on v5, as this change might require us to change minimum supported version for Moq. Unless somebody wants to look upfront to prepare Moq for this.

I'll update this issue once we start work on v5.

**Comment by riezebosch** [comment]
```csharp
    public class ValueProviderTests
    {
        [Fact]
        public void UseValueProvider()
        {
            var fixture = new Fixture();
            var mock = new Mock<IInterfaceWithGenericMethod>
            {
                DefaultValueProvider = new AutoFixtureValueProvider(fixture)
            };


            Assert.NotNull(mock.Object.GenericMethod<string>());
        }
        
        [Fact]
        public void CustomizedAnonymousType()
        {
            var fixture = new Fixture();
            fixture.Customize<string>(x => x.FromSeed(y => "asf"));
            var mock = new Mock<IInterfaceWithGenericMethod>
            {
                DefaultValueProvider = new AutoFixtureValueProvider(fixture)
            };


            Assert.Equal("asf", mock.Object.GenericMethod<string>());
        }

        public class AutoFixtureValueProvider : DefaultValueProvider
        {
            private readonly IFixture _fixture;

            public AutoFixtureValueProvider(IFixture fixture)
            {
                _fixture = fixture;
            }

            protected override object GetDefaultValue(Type type, Mock mock) =>
                _fixture.Create(type, new SpecimenContext(_fixture));
        }
    }
```

This is kind of impressive, I guess.

**Comment by stakx** [comment]
@riezebosch: Looking good. Btw., if you subclass `LookupOrFallbackDefaultValueProvider` instead of `DefaultValueProvider`, you'll get automatic support for default values wrapped in completed `Task<>` and `ValueTask<>`. In case you want that.

**Comment by stakx** [comment]
@zvirja: Thanks for the reminder!

Someone noticed a while ago over at Moq's repo that due to the accessibility of `DefaultValueProvider`'s methods, these providers aren't as composable as they could be. I've been wanting to fix this in one of the next minor releases. It'll likely be a very small breaking change. I thought I'd mention it here so that we can coordinate versions / release schedules if needed.

Also, if anyone here finds other problems with Moq's `DefaultValueProvider`, please let me know.

**Comment by riezebosch** [comment]
Checkout my work on the value provider: https://github.com/riezebosch/AutoFixture/tree/value-provider. Rewrote the tests a bit that explicitly create a `Mock<T>` from the fixture and assert on `.Object`

What's not working (yet): 
1. ref/out parameters, moq doesn't seem to visit the value provider for that?!
1. delegates
1. removed the configure members (for now).

**Comment by stakx** [comment]
> ref/out parameters, moq doesn't seem to visit the value provider for that 

As far as I can tell, from Moq's perspective, that's by design:

* `ref` parameters are assumed to already have a value assigned before the method call, so there's no strict need to set them.

* `out` parameters are historically treated in a very peculiar way: they're *read* at setup time to determine the desired value that the argument should be set to at the end of each method call.

That's why the default value provider doesn't get queried for those by-refs.

If AutoFixture always sets by-ref parameters, then Moq and AutoFixture have different modes of operation, and I'm not yet sure how we can seamlessly bring the two together.

> Had to use reflection to invoke the SetupAllProperties since it's only public available on `Mock<T>`

Yes, Moq doesn't have a non-generic public API, it's in the planning however (see e.g. https://github.com/moq/moq4/issues/887). For the moment, I'm afraid you'll need that reflection detour.

**Comment by riezebosch** [comment]
> > ref/out parameters, moq doesn't seem to visit the value provider for that
> 
> As far as I can tell, from Moq's perspective, that's by design:
> 
> * `ref` parameters are assumed to already have a value assigned before the method call, so there's no strict need to set them.
> * `out` parameters are historically treated in a very peculiar way: they're _read_ at setup time to determine the desired value that the argument should be set to at the end of each method call.
> 
> That's why the default value provider doesn't get queried for those by-refs.
> 
> If AutoFixture always sets by-ref parameters, then Moq and AutoFixture have different modes of operation, and I'm not yet sure how we can seamlessly bring the two together.
> 
> > Had to use reflection to invoke the SetupAllProperties since it's only public available on `Mock<T>`
> 
> Yes, Moq doesn't have a non-generic public API, it's in the planning however (see e.g. [moq/moq4#887](https://github.com/moq/moq4/issues/887)). For the moment, I'm afraid you'll need that reflection detour.

Minor correction, for the `ref` parameter the value provider _is_ visited but for some reason the variable doesn't get overwritten (probably boxing or something like that?). For the `out` parameter the provider is not visited at all.

For some reason the ref parameters were ignored in current implementation so I guess this is already a step forward.

**Comment by stakx** [comment]
@riezebosch:

> Minor correction, for the ref parameter the value provider is visited but for some reason the variable doesn't get overwritten (probably boxing or something like that?). 

Strange... that shouldn't happen. May I ask you to post a small demo code&mdash;either here or (perhaps better, so we don't hijack this issue) over at the moq/moq4 repo?

**Comment by riezebosch** [comment]
It's in my reply above: https://github.com/AutoFixture/AutoFixture/issues/1139#issuecomment-534487978

**Comment by stakx** [comment]
> It's in my reply above: #1139 (comment)

I ran those tests, which fail (as expected).

> Minor correction, for the ref parameter the value provider is visited 

I tried to reproduce that claim, but failed. I changed your `TestDefaultValueProvider` as follows:

```diff
 private class TestValueProvider : DefaultValueProvider
 {
     private readonly object _result;
     public TestValueProvider(object result) => _result = result;
-    protected override object GetDefaultValue(Type type, Mock mock) => _result;
+    protected override object GetDefaultValue(Type type, Mock mock) => throw new InvalidOperationException("unexpectedly invoked default value provider");
 }
```

However, that exception never gets thrown, meaning `GetDefaultValue` doesn't get called by the tests. What am I missing?

**Comment by riezebosch** [comment]
My guess is the exception gets swallowed by Moq. I slightly modified the provider:

```csharp
[Fact]
public void RefParameter()
{
    var provider = new TestValueProvider("bye");
    var mock = new Mock<IInterfaceWithRefMethod> { DefaultValueProvider = provider };
    var s = "hi";
    mock.Object.Method(ref s);
    
    Assert.True(provider.IsInvoked);
    Assert.Equal("bye", s);
}
[Fact]
public void OutParameter()
{
    var provider = new TestValueProvider(4);
    var mock = new Mock<IInterfaceWithOutMethod> { DefaultValueProvider = provider };
    mock.Object.Method(out var s);
    
    Assert.True(provider.IsInvoked);
    Assert.Equal(4, s);
}
private class TestValueProvider : DefaultValueProvider
{
    private readonly object _result;
    public TestValueProvider(object result) => _result = result;
    protected override object GetDefaultValue(Type type, Mock mock)
    {
        IsInvoked = true;
        return _result;
    }
    public bool IsInvoked { get; set; }
}
```

**Comment by stakx** [comment]
@riezebosch, I cannot reproduce your results, using Moq 4.13.0, your above code, and the two interfaces as defined below. Both tests fail due to the default value provider not having been invoked (as expected).

```csharp
public interface IInterfaceWithRefMethod { void Method(ref string arg); }
public interface IInterfaceWithOutMethod { void Method(out int arg);    }
```

**Comment by riezebosch** [comment]
You're right, the value provider is visited for the ref parameter method but only because the interface method is defined as `string Method(ref string s);`.
